lib/Runtime/ByteCode/ByteCodeGenerator.cpp

```diff
@@ -3338,17 +3340,26 @@ void PostVisitBlock(ParseNode *pnode, ByteCodeGenerator *byteCodeGenerator)
         return;
     }
 
+    Scope *scope = pnode->sxBlock.scope;
     if (pnode->sxBlock.GetCallsEval() || pnode->sxBlock.GetChildCallsEval() || (byteCodeGenerator->GetFlags() & (fscrEval | fscrImplicitThis | fscrImplicitParents)))
     {
-        Scope *scope = pnode->sxBlock.scope;
         bool scopeIsEmpty = scope->IsEmpty();
         scope->SetIsObject();
         scope->SetCapturesAll(true);
         scope->SetMustInstantiate(!scopeIsEmpty);
     }
 
-    byteCodeGenerator->PopScope();
-    byteCodeGenerator->PopBlock();
+    if (scope->GetScopeType() != ScopeType_Parameter)
+    {
+        byteCodeGenerator->PopScope();
+        byteCodeGenerator->PopBlock();
+    }
+    else
+    {
+        // If current block is a param scope then it was not pushed in during PreVisitBlock. So let's not do a pop here.
```

Comment:
Nope. It’ll fall out of the way we instantiate them. A formal captured by a closure in a default expression isn’t accessible from within the body, regardless of whether the body declares it locally. So the instantiation is effectively the same whether we consider them nested or not.

--Paul

From: Ian Halliday [mailto:notifications@github.com]
Sent: Wednesday, February 10, 2016 6:52 PM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>
Subject: Re: [ChakraCore] Enabling split scope in lambda and adding more test cases (#289)


In lib/Runtime/ByteCode/ByteCodeGenerator.cpp<https://github.com/Microsoft/ChakraCore/pull/289#discussion_r52559409>:

>          bool scopeIsEmpty = scope->IsEmpty();

>          scope->SetIsObject();

>          scope->SetCapturesAll(true);

>          scope->SetMustInstantiate(!scopeIsEmpty);

>      }

>

> -    byteCodeGenerator->PopScope();

> -    byteCodeGenerator->PopBlock();

> +    if (scope->GetScopeType() != ScopeType_Parameter)

> +    {

> +        byteCodeGenerator->PopScope();

> +        byteCodeGenerator->PopBlock();

> +    }

> +    else

> +    {

> +        // If current block is a param scope then it was not pushed in during PreVisitBlock. So let's not do a pop here.

What now? The spec calls for the body scope to be nested in the parameter scope. Why do we want to make them siblings? Won't that require special casing look-ups everywhere?

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/289/files#r52559409>.